File: examples/sdk/typescript/04-actions.ts
/**
 * Example 04: Actions — execute actions and track results
 *
 * Demonstrates: Action discovery, execution, async polling, execution logs.
 *
 * Run: npx tsx examples/sdk/04-actions.ts
 */
import { createClient, findKnWithData, pp } from "./setup.js";

async function main() {
  const client = await createClient();
  const { knId, knName } = await findKnWithData(client);
  console.log(`Using BKN: ${knName} (${knId})\n`);

  // 1. List available action types
  const actionTypes = await client.knowledgeNetworks.listActionTypes(knId);
  const ats = actionTypes as Array<{ id?: string; name?: string; description?: string }>;
  console.log(`=== Action Types (${ats.length}) ===`);
  for (const at of ats) {
    console.log(`  ${at.name} (${at.id}) — ${at.description ?? ""}`);
  }

  if (ats.length === 0) {
    console.log("\nNo action types found. This BKN has no executable actions.");
    return;
  }

  // 2. Query an action type to see its schema/parameters
  const at = ats[0];
  console.log(`\n=== Action Detail: "${at.name}" ===`);
  try {
    const actionDetail = await client.bkn.queryAction(knId, at.id!, {});
    pp(actionDetail);
  } catch (e) {
    console.log(`  (query failed — action's backing datasource may be unavailable: ${(e as Error).message})`);
  }

  // 3. List action execution logs (historical runs)
  console.log("\n=== Action Logs ===");
  try {
    const logs = await client.bkn.listActionLogs(knId, { atId: at.id, limit: 5 });
    console.log(`Found ${logs.length} log(s) for "${at.name}":`);
    for (const log of logs as Array<{ id?: string; status?: string; created_at?: string }>) {
      console.log(`  [${log.status}] ${log.id} — ${log.created_at ?? ""}`);
    }

    // 4. Get detail of the most recent log (if any)
    if (logs.length > 0) {
      const firstLog = logs[0] as { id?: string };
      if (firstLog.id) {
        console.log(`\n=== Log Detail: ${firstLog.id} ===`);
        const detail = await client.bkn.getActionLog(knId, firstLog.id);
        pp(detail);
      }
    }
  } catch (e) {
    console.log(`  (logs unavailable — execution index may not exist yet: ${(e as Error).message})`);
  }

  // =====================================================================
  // UNCOMMENT BELOW TO EXECUTE (WRITE OPERATION — triggers real side effects)
  // =====================================================================
  //
  // const execution = await client.bkn.executeAction(knId, at.id!, { params: {} });
  // console.log("Execution started:", execution);
  //
  // // Poll until complete
  // const execResult = execution as { execution_id?: string };
  // if (execResult.execution_id) {
  //   let status: unknown;
  //   do {
  //     await new Promise((r) => setTimeout(r, 2000));
  //     status = await client.bkn.getExecution(knId, execResult.execution_id);
  //     console.log("Status:", (status as { status?: string }).status);
  //   } while ((status as { status?: string }).status === "running");
  //   console.log("Final result:");
  //   pp(status);
  // }
}

main().catch(console.error);

File: examples/sdk/python/04-actions.py
"""Example 04: Actions — execute actions and track results.

Demonstrates: action discovery, execution, async polling, execution logs.

Run: python examples/sdk/04-actions.py
"""

from __future__ import annotations

from setup import create_client, find_kn_with_data, pp


def main() -> None:
    client = create_client()
    kn_id, kn_name = find_kn_with_data(client)
    print(f"Using BKN: {kn_name} ({kn_id})\n")

    action_types = client.action_types.list(kn_id)
    print(f"=== Action Types ({len(action_types)}) ===")
    for at in action_types:
        print(f"  {at.get('name')} ({at.get('id')}) — {at.get('description', '')}")

    if not action_types:
        print("\nNo action types found. This BKN has no executable actions.")
        return

    at = action_types[0]
    at_id = at.get("id")
    print(f'\n=== Action Detail: "{at.get("name")}" ===')
    try:
        action_detail = client.action_types.query(kn_id, at_id, {})
        pp(action_detail)
    except Exception as e:
        print(f"  (query failed — action's backing datasource may be unavailable: {e})")

    print("\n=== Action Logs ===")
    try:
        # The Python list_logs() doesn't filter by action_type_id today —
        # we list all recent logs and filter client-side for parity with TS.
        all_logs = client.action_types.list_logs(kn_id, limit=20)
        logs = [log for log in all_logs if str(log.get("at_id") or log.get("action_type_id") or "") == at_id][:5]
        print(f'Found {len(logs)} log(s) for "{at.get("name")}":')
        for log in logs:
            print(f"  [{log.get('status')}] {log.get('id')} — {log.get('created_at', '')}")

        if logs and logs[0].get("id"):
            first_id = logs[0]["id"]
            print(f"\n=== Log Detail: {first_id} ===")
            detail = client.action_types.get_log(kn_id, first_id)
            pp(detail)
    except Exception as e:
        print(f"  (logs unavailable — execution index may not exist yet: {e})")

    # =====================================================================
    # UNCOMMENT BELOW TO EXECUTE (WRITE OPERATION — triggers real side effects)
    # =====================================================================
    #
    # execution = client.action_types.execute(kn_id, at_id, {})
    # print("Execution started:", execution)
    # # action_types.execute() already polls until completion; result is the final state.


if __name__ == "__main__":
    main()
